src/libPMacc/include/traits/GetUniqueTypeId.hpp

```diff
@@ -57,27 +60,47 @@ const uint64_t GetUniqueTypeId<T_Type>::id = uint64_t(&GetUniqueTypeId<T_Type>::
 
 /** Get a unique id of a type
  *
- * - generate a unique 64bit id of a type at *runtime*
+ * - generate a unique id of a type at *runtime*
  * - the id of a type is equal on each instance of a process
  *
  * @tparam T_Type any object (class or typename)
- *
- * @treturn ::uid
+ * @tparam T_ResultType result type
  */
-template<typename T_Type>
+template<typename T_Type, typename T_ResultType = uint64_t>
 struct GetUniqueTypeId
 {
-    static const uint64_t uid;
-};
 
-/** instantiation of traits::GetUniqueTypeId
- *
- * - create a instance of `traits::GetUniqueTypeId` and initialize the uid
- * - map `detail::GetUniqueTypeId<T>` to a range [1; 2^64-1]
- *   `uid = unique_id_of_current_type - base_unique_id`
- */
-template<typename T_Type>
-const uint64_t GetUniqueTypeId<T_Type>::uid = detail::GetUniqueTypeId<T_Type>::id - detail::GetUniqueTypeId<uint8_t>::id;
+    /** create unique id
+     *
+     * @param maxValue largest allowed id
+     */
+    static const T_ResultType uid(uint64_t maxValue = uint64_t(1) << (sizeof (T_ResultType) * 8) - uint64_t(1))
```

Comment:
I can't find a source that clearly says that CHAR_BIT is always 8. 
http://www.cplusplus.com/reference/climits/ said that it is >= 8.

On 27.07.2015 14:23, Benjamin Worpitz wrote:
> In src/libPMacc/include/traits/GetUniqueTypeId.hpp
> <https://github.com/ComputationalRadiationPhysics/picongpu/pull/962#discussion_r35529598>:
>
>>
>> -/** instantiation of traits::GetUniqueTypeId
>> - *
>> - * - create a instance of `traits::GetUniqueTypeId` and initialize the uid
>> - * - map `detail::GetUniqueTypeId<T>` to a range [1; 2^64-1]
>> - *   `uid = unique_id_of_current_type - base_unique_id`
>> - */
>> -template<typename T_Type>
>> -const uint64_t GetUniqueTypeId<T_Type>::uid = detail::GetUniqueTypeId<T_Type>::id - detail::GetUniqueTypeId<uint8_t>::id;
>> +    /** create unique id
>> +     *
>> +     * @param maxValue largest allowed id
>> +     */
>> +    static const T_ResultType uid(uint64_t maxValue = uint64_t(1) << (sizeof (T_ResultType) * 8) - uint64_t(1))
>
> No, that is not true. In C and C++ a byte is defined to be equal to
> |char| and must have at least 8 bist. |CHAR_BIT| defines the real number
> of bits per byte.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/ComputationalRadiationPhysics/picongpu/pull/962/files#r35529598>.
>

-- 
René Widera
Abteilung Laser-Teilchenbeschleunigung (FWKT)
Helmholtz-Zentrum Dresden-Rossendorf
Tel: +49 (0351) 260 3543
r.widera@hzdr.de
http://www.hzdr.de

Vorstand: Prof. Dr. Dr. h. c. Roland Sauerbrey,
           Prof. Dr. Dr. h. c. Peter Joehnk
Vereinsregister: VR 1693 beim Amtsgericht Dresden